aea/helpers/exec_timeout.py

```diff
@@ -0,0 +1,154 @@
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+#
+# ------------------------------------------------------------------------------
+
+"""Python code execution time limit tools."""
+
+import asyncio
+import ctypes
+import signal
+import threading
+from abc import ABC, abstractmethod
+from asyncio import Future
+from asyncio.events import AbstractEventLoop
+from types import TracebackType
+from typing import Type
+
+
+class TimeoutResult:
+    """ Result of ExecTimeout context manager """
+
+    def __init__(self):
+        self.cancelled_by_timeout = False
+
+    def set_cancelled_by_timeout(self):
+        self.cancelled_by_timeout = True
+
+    def is_cancelled_by_timeout(self):
+        return self.cancelled_by_timeout
+
+
+class TimeoutException(Exception):
+    pass
+
+
+class BaseExecTimeout(ABC):
+    def __init__(self, timeout: float = 0):
+        self.timeout = timeout
+        self.result = TimeoutResult()
+
+    def _on_timeout(self, *args, **kwargs):
+        raise TimeoutException()
+
+    def __enter__(self):
+        if self.timeout:
+            self._set_timeout_watch()
+
+        return self.result
+
+    def __exit__(
+        self, exc_type: Type[Exception], exc_val: Exception, exc_tb: TracebackType
+    ) -> bool:
+        if self.timeout:
+            self._remove_timeout_watch()
+
+        if isinstance(exc_val, TimeoutException):
+            self.result.set_cancelled_by_timeout()
+            return True
+        return False
+
+    @abstractmethod
+    def _set_timeout_watch(self):
+        raise NotImplementedError  # pragma: nocover
+
+    @abstractmethod
+    def _remove_timeout_watch(self):
+        raise NotImplementedError  # pragma: nocover
+
+
+class ExecTimeoutSigAlarm(BaseExecTimeout):
+    """ use only in main thread """
+
+    def _set_timeout_watch(self):
+        signal.setitimer(signal.ITIMER_REAL, self.timeout, 0)
+        signal.signal(signal.SIGALRM, self._on_timeout)
+
+    def _remove_timeout_watch(self):
+        signal.setitimer(signal.ITIMER_REAL, 0, 0)
+
+
+class ExecTimeoutThreadGuard(BaseExecTimeout):
+    exception_class = TimeoutException  # type: Type[Exception]
+
+    _supervisor_thread = None  # type: threading.Thread
+    _loop = None  # type: AbstractEventLoop
+    _stopped_future = None  # type: asyncio.Future
+
+    @classmethod
+    def start(cls) -> None:
+        """
+        start supervisor thread to check timeouts
+
+        :return: None
+        """
+        if cls._supervisor_thread:
+            return
+
+        cls._loop = asyncio.new_event_loop()
```

Comment:
Does it need to be a *new* event loop?

Reply:
I think so, cause it started in dedicated thread. It's better to have own event loop just for code execution time control and do not rely on other loops present in the process.